File: design-pattern/observer-pattern/observer.js
;
;
var Channel = /** @class */ (function () {
    function Channel(name) {
        this.subscriberList = [];
        this.latestVideo = "";
        this.name = "";
        this.name = name;
    }
    ;
    Channel.prototype.subscribe = function (subscribe) {
        this.subscriberList.push(subscribe);
    };
    ;
    Channel.prototype.unsubscribe = function (subscribe) {
        this.subscriberList.filter(function (i) { return i != subscribe; });
    };
    ;
    Channel.prototype.notify = function (video) {
        this.latestVideo = video;
        console.log("new video uploaded by" + this.name);
        this.subscriberList.forEach(function (i) { return i.update(); });
    };
    Channel.prototype.getLatestVideo = function () {
        return this.latestVideo;
    };
    ;
    return Channel;
}());
var Subscriber = /** @class */ (function () {
    function Subscriber(name, channelcurr) {
        this.name = "";
        this.name = name;
        this.subscribedChannel = channelcurr;
    }
    ;
    Subscriber.prototype.update = function () {
        this.subscribedChannel.getLatestVideo();
    };
    ;
    return Subscriber;
}());
;
var omkarsonavale = new Channel("omkarfun");
var nickolodean = new Channel("nickolodean");
var user1 = new Subscriber("user1", omkarsonavale);
var user2 = new Subscriber("user2", omkarsonavale);
var user3 = new Subscriber("user3", omkarsonavale);
var user4 = new Subscriber("user4", omkarsonavale);
omkarsonavale.notify("i am uploading s3 video");

File: design-pattern/observer-pattern/observer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Isubscriber
{
public:
    virtual void update() = 0;
};

class Ichannel
{
public:
    virtual void subscribe(Isubscriber *isub) = 0;
    virtual void unsubscribe(Isubscriber *isub) = 0;
    virtual void notify() = 0;
    virtual ~Ichannel() {};
};

class Channel : public Ichannel
{
private:
    vector<Isubscriber *> subscribers;
    string name;
    string videoTitle;

public:
    Channel(string name)
    {
        this->name = name;
    }

    void subscribe(Isubscriber *isub) override
    {
        subscribers.push_back(isub);
    }

    void unsubscribe(Isubscriber *isub) override
    {
        auto it = find(subscribers.begin(), subscribers.end(), isub);
        if (it != subscribers.end())
        {
            subscribers.erase(it);
        }
    };

    void notify() override
    {
        for (auto sub : subscribers)
        {
            sub->update();
        }
    }

    void uploadVideo(string title)
    {
        videoTitle = title;
        cout << "new video realeses " << name << endl;
        this->notify();
    };

    string getNameOfUpload()
    {
        return this->videoTitle;
    }
};

class Subscriber : public Isubscriber
{
private:
    Channel *channel1;
    string name;

public:
    Subscriber(Channel *c1, string name)
    {
        this->channel1 = c1;
        this->name = name;
    }

    void update() override
    {
        cout << channel1->getNameOfUpload() << endl;
    }
};

int main(void)
{
    Channel *c1 = new Channel("omkar13");
    Channel *c2 = new Channel("mukesh12");

    Subscriber *s1 = new Subscriber(c1, "omkar");
    Subscriber *s2 = new Subscriber(c1, "omkar1");
    Subscriber *s3 = new Subscriber(c1, "omkar2");

    c1->subscribe(s1);
    c1->subscribe(s2);
    c1->subscribe(s3);

    c1->uploadVideo("this is my tour video");
};